### ISSUE TITLE
[BUG] "@latest" vs "@master", the "latest" wins

### ISSUE THREAD

**Issue by KEINOS** [nao1215/gup#64]
## About the problem

Packages installed via "`go install <importPath>@master`" or "`go install <importPath>@main`" gets replaced into "`<importPath>@latest`" after `gup update`.

It seems similar to #36 but a little different. Since the down-grading issue has been fixed.

- Related issue:
  - #36

## Steps to reproduce

```shellsession
$ # Install the latest master (not the latest release)
$ go install github.com/msoap/go-carpet@master
** snip **

$ go-carpet -version
1.9.0

$ # -mincov option exists
$ go-carpet -h 2>&1 | grep mincov
  -mincov float

$ gup update
** snip **

$ go-carpet -version
1.9.0

$ # -mincov option does not exist
$ go-carpet -h 2>&1 | grep mincov
$
```

> __Note__
>
> This behavior does not happen if the packages are never version-tagged. Because in that case the versions are `v0.0.0-<commit id>`.
>
> This occurs when the package owner has not published/released the changes made to the current `master` or `main` branch.

## Expected behavior

If the installed packages are tagged such as `@master` or `@main` then `gup update` should install the latest source and not the latest release.

## Possible causes

1. Even if the previous and next versions are the same, `gup update` will install the "latest" packages.

https://github.com/nao1215/gup/blob/763171a400208f29cdb1e24e974fa70b02bb1daf/cmd/update.go#L75-L140

https://github.com/nao1215/gup/blob/763171a400208f29cdb1e24e974fa70b02bb1daf/internal/goutil/goutil.go#L189-L204

## Possible fixes

This is a difficult task. The reason is that it depends on how the term "`update`" is treated.

My current proposes are:

1. Compare the versions and skip the [update](https://github.com/nao1215/gup/blob/763171a400208f29cdb1e24e974fa70b02bb1daf/cmd/update.go#L75) function if they are the same. (`-f, --force` flag for the original behavior)
    - PROS: Simple
    - CONS: Gets slow (due to the [GetLatestVer](https://github.com/nao1215/gup/blob/763171a400208f29cdb1e24e974fa70b02bb1daf/internal/goutil/goutil.go#L207-L213) call) and `@master` installed app never gets updated unless re-installing manually.
2. Obtain the version tag from the pulled repo and use it if they are not `@v??.??.??` or `@latest`  (e.g. `@master` or `@main`), otherwise use `@latest`.
    - PROS: It will follow the initial `go install <package>@<branch>` installation.
    - CONS: Don't know how to obtain the version tag ...

## Current workaround

- Reinstall manually via `go install <package>@<branch>` until the package owner comes back from his/her vacation.

## Env info

- OS: macOS 12.6.1 (OSX Monterey) and Alpine Linux v3.16 (via Docker, `golang:alpine` image)
- Go Version: 1.19.3 (Both macOS and Linux)
- Application Version: gup v0.15.1

**Comment by nao1215** [comment]
@KEINOS 

Thank you for your bug report. It's very detailed.
I prefer your suggestion 2.
>Obtain the version tag from the pulled repo and use it if they are not @v??.??.?? or @latest (e.g. @master or @main), otherwise use @latest.

However, I do not know how to fix it, so I cannot fix the code right away.
Furthermore, I would like to spend some time on another project ([sqly](https://github.com/nao1215/sqly)) for a while. I want to hurry up and complete the implementation of sqly command because I use the sqly command in my work.

So, I will be a little late in fixing  gup command.

P.S.
I see that you frequently use the gup command (⌒▽⌒) Thank you

**Comment by KEINOS** [comment]
@nao1215 

> I would like to spend some time on another project ([sqly](https://github.com/nao1215/sqly)) for a while.

Cool project! 👍 Take your time as much as you want.
Don't worry. it's not critical and we have a workaround so far.

> However, I do not know how to fix it, so I cannot fix the code right away.

Me too.😄

Though, digging the `go install` behavior, I have a feeling we have to take 2 steps such as `gup update` then `gup upgrade`, like the other package-managers do.

If that's so, we should apply them in the next major update.

P.S.
Yes. `gup` is one of my "must install" tools!

**Comment by nao1215** [comment]
I don't see how gup can automatically and fast determine whether @latest or @master should be adopted.

### Another idea
-  Add --master option for "gup update"
When --master is enabled, the "go install" update will use @master as the version identifier.

- Record which binaries were updated with "go install XXX@master" in the file.
If the --master option is used, ask the following question. "In future updates, do you want to update the binary you just updated as @master? (Do you want to avoid using @latest?)"
If the user answers "Yes", then the binaries to be updated with @master in the future will be recorded in the configuration file.

- When "gup update", gup checks the configuration file to determine whether to update with @master or @latest

**Comment by nao1215** [comment]
@KEINOS 

It has been a long time. I have found the solution.
If we "go install XXX@main" or "go install XXX@master" golang binary, we will get the following version.
This version includes build metadata.
```
[major].[minor].[patch]-[pre-release].[build metadata]
```
The important thing here is that there is a way to get build metadata.

For example
```
    buildInfo, err := buildinfo.ReadFile("binary_name")
    if err != nil {
        return
    }

    if buildInfo.Main.Sum != "" {
        // install by @main or @master
        fmt.Println("Build metadata included")
    } else {
        // @latest install
        fmt.Println("Build metadata not included")
    }
```

We can use the above code to switch between @latest and @master (or @main).

**Comment by KEINOS** [comment]
@nao1215 

That's a huge progress!

I confirm that installing with `@master` or `@main` tag contains more version information. Such as `pre-release` and `build metadata`.

```shellsession
$ # Install specific version
$ go install github.com/nao1215/gup@v0.15.1
** snip **

$ gup version
gup version v0.15.1 (under Apache License version 2.0)

$ which gup
/go/bin/gup

$ go run main.go "$(which gup)"
Build metadata included
debug.Module{
  Path: "github.com/nao1215/gup",
  Version: "v0.15.1",
  Sum: "h1:52nuQnjIJ+KF3MJX5NhSndR41Fl7ruTzHUKI6dmVHaE=",
  Replace: (*debug.Module)(nil),
}
```

```shellsession
$ # @latest
$ go install github.com/nao1215/gup@latest
** snip **

$ gup version
gup version v0.17.0 (under Apache License version 2.0)

$ which gup
/go/bin/gup

$ go run main.go "$(which gup)"
Build metadata included
debug.Module{
  Path: "github.com/nao1215/gup",
  Version: "v0.17.0",
  Sum: "h1:eY7VWCBJ28dCd4Y23HzBKpCrjmOOereUzafnQakjnXc=",
  Replace: (*debug.Module)(nil),
}
```

```shellsession
$ # @main
$ go install github.com/nao1215/gup@main
** snip **

$ gup version
gup version v0.17.1-0.20230220114040-360e9b88f2fd (under Apache License version 2.0)

$ which gup
/go/bin/gup

$ go run main.go "$(which gup)"
Build metadata included
debug.Module{
  Path: "github.com/nao1215/gup",
  Version: "v0.17.1-0.20230220114040-360e9b88f2fd",
  Sum: "h1:gC7kNqzmJ3lG1MQRCO0E1N0K5iIW/YDhXhJXKdEQ0nc=",
  Replace: (*debug.Module)(nil),
}
```

<details><summary>main.go</summary>

```go
package main

import (
	"debug/buildinfo"
	"fmt"
	"log"
	"os"

	"github.com/Code-Hex/dd"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run main.go <path_to_binary>")
		return
	}

	buildInfo, err := buildinfo.ReadFile(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}

	if buildInfo.Main.Sum != "" {
		// install by @main or @master
		fmt.Println("Build metadata included")
		fmt.Println(dd.Dump(buildInfo.Main))
	} else {
		// @latest install
		fmt.Println("Build metadata not included")
		fmt.Println(dd.Dump(buildInfo.Main))
	}
}
```

</details>

Though, as you can see the above, in any cases (`@latest`, `@main`, etc.) the `buildInfo.Main.Sum` contains the hash. Thus, all of them returns "Build metadata included".

I think we should check the "build" from the "build-metadata" in the version using [semver](https://pkg.go.dev/golang.org/x/mod/semver) package.

```go
package main

import (
	"fmt"
	"strings"

	"golang.org/x/mod/semver"
)

func Example() {
	for _, tt := range []string{
		"v0.17.0-alpha",                         // @latest (alpha release)
		"v0.17.0",                               // @latest
		"v0.17.1-0.20230220114040-360e9b88f2fd", // @main/@master
	} {
		fmt.Printf(
			"%s --> Has build ver: %v (Build:%s)\n",
			tt, HasBuildVersion(tt), GetVersionBuild(tt),
		)
	}
	// Output:
	// v0.17.0-alpha --> Has build ver: false (Build:)
	// v0.17.0 --> Has build ver: false (Build:)
	// v0.17.1-0.20230220114040-360e9b88f2fd --> Has build ver: true (Build:+360e9b88f2fd)
}

// HasBuildVersion returns true if the given version info contains a build hash.
func HasBuildVersion(ver string) bool {
	return GetVersionBuild(ver) != ""
}

// GetVersionBuild returns the build version from the git tagged versioning.
func GetVersionBuild(ver string) string {
	// Convert git tag's format to Go semver style.
	//   e.g.) v1.2.3-4-5678 --> v1.2.3-4+5678
	if preCount := strings.Count(ver, "-"); preCount > 1 {
		ver = strings.ReplaceAll(ver, "-", "+")
		ver = strings.Replace(ver, "+", "-", preCount-1)
	}

	return semver.Build(ver)
}
```

- https://go.dev/play/p/S3ShYnY2GsA @ GoPlayground

**Comment by nao1215** [comment]
@KEINOS 
Thanks for researching the build-info.
I will fix this bug. So, I find time to fix it.

If I create a PR, I will contact you before the merge.

**Comment by KEINOS** [comment]
> If I create a PR, I will contact you before the merge.

(￣^￣)ゞ

**Comment by nao1215** [comment]
Related PR #91

**Comment by KEINOS** [comment]
I confirm that "[Update binaries with @main or @master](https://github.com/nao1215/gup#update-binaries-with-main-or-master)" fixes this issue (my problem).

So, closing. Thank you!

---

BTW, to make `gup` more flexible I prefer the arguments as below:

```diff
- $ gup update --main=gup,lazygit,sqly
+ $ gup update --branch=main gup lazygit sqly

- $ gup update --master=gup,lazygit,sqly
+ $ gup update --branch=master gup lazygit sqly
```

Though, this approach is a bit niche. And **the current implementation is enough for me**, so feel it as "just-an-idea".

Again, thank you for the fix! I am very happy about the fix! 👍

**Comment by nao1215** [comment]
@KEINOS 
PR #91 was intended as a temporary solution. I had been thinking about introducing a better solution. However, nearly a year has passed. The method I had previously considered was somewhat challenging to implement. Therefore, I introduced PR #91 as a stopgap measure.

gup is the most popular among the open-source projects I've created, so I'm always thinking about introducing better improvements.

Thank you for your new proposal.
At some point when I have time, I will resume updates for gup.
